src/Shared/Buffers/RefPooledArrayBufferWriter.cs

```diff
@@ -0,0 +1,266 @@
+    /// <exception cref="InvalidOperationException">Thrown if advancing would exceed buffer capacity.</exception>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void Advance(int count)
+    {
+        ThrowIfDisposed();
+        if (count < 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(count), "Count cannot be negative.");
+        }
+
+        var canAdvance = _rentedBuffer is not null
+            ? _index + count < _rentedBuffer.Length
+            : _index + count < _buffer.Length;
+
+        if (!canAdvance)
+        {
+            throw new InvalidOperationException($"Cannot advance past the end of the buffer. Current position: {_index}, Capacity: {_buffer.Length}, Requested advance: {count}.");
+        }
+        _index += count;
+    }
+
+    private void CheckAndResizeBuffer(int sizeHint)
+    {
+        if (sizeHint <= 0)
+        {
+            throw new ArgumentOutOfRangeException(nameof(sizeHint), actualValue: sizeHint, $"{nameof(sizeHint)} ('{sizeHint}') must be a non-negative value.");
+        }
+        if (sizeHint == 0)
+        {
+            sizeHint = MinimumBufferSize;
+        }
+
+        if (_rentedBuffer is null)
+        {
+            var bufferSpace = _buffer.Length - _index;
+            if (bufferSpace < sizeHint)
+            {
+                // initial buffer is not enough, we need to start renting from the pool
+                var rentedInitialSize = _buffer.Length + Math.Max(sizeHint, _buffer.Length);
+                _rentedBuffer = ArrayPool<T>.Shared.Rent(rentedInitialSize);
+
+                _buffer.CopyTo(_rentedBuffer);
+                _buffer.Clear();
+
+                Debug.Assert(_rentedBuffer.Length - _index > 0);
+                Debug.Assert(_rentedBuffer.Length - _index >= sizeHint);
+            }
+
+            return;
+        }
+
+        // we are already using rented buffer, so grow it if needed
+        var availableSpace = _rentedBuffer.Length - _index;
+        if (sizeHint > availableSpace)
+        {
+            var growBy = Math.Max(sizeHint, _rentedBuffer.Length);
+            var newSize = checked(_rentedBuffer.Length + growBy);
+
+            var oldBuffer = _rentedBuffer;
+            _rentedBuffer = ArrayPool<T>.Shared.Rent(newSize);
+
+            Debug.Assert(oldBuffer.Length >= _index);
+            Debug.Assert(_rentedBuffer.Length >= _index);
+
+            var previousBuffer = oldBuffer.AsSpan(0, _index);
+            previousBuffer.CopyTo(_rentedBuffer);
+            previousBuffer.Clear();
+            ArrayPool<T>.Shared.Return(oldBuffer);
+        }
+        Debug.Assert(_rentedBuffer.Length - _index > 0);
+        Debug.Assert(_rentedBuffer.Length - _index >= sizeHint);
+    }
+
+    /// <summary>
+    /// Clears the buffer, resetting the write position to zero without deallocating.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public void Clear()
```

Comment:
What calls this? Even if we need this, I probably wouldn't bother doing more than resetting the index.

Reply:
nobody. I just tried to do the same API as exists in `class PooledArrayBufferWriter<T>`. I think we should be good to delete any unused API, and return if needed.